libexec/bats-core/bats

```diff
@@ -363,13 +464,23 @@ for filename in "${arguments[@]}"; do
 
   if [[ -d "$filename" ]]; then
     shopt -s nullglob
-    if [[ "$recursive" -eq 1 ]]; then
+    if [[ "$recursive" == "0" ]]; then
+      build_find_command_string "find_command"
+
+      # shellcheck disable=SC2154 
       while IFS= read -r -d $'\0' file; do
-        filenames+=("$file")
-      done < <(find -L "$filename" -type f -name "*.${BATS_FILE_EXTENSION:-bats}" -print0 | sort -z)
+        filenames+=("$file")   
+        # performance boost through piping by /dev/fd/63 instead of
+        # creating a temporary file for find's results.
+        # => pipe [<(..)] and redirect [<] '$find_command' to standard input.
+        # for `read` using c-style string termination '\0' [-print0]
+        # eval is a built-in command, no performance issue's expected
+      done < <(eval "$find_command" -print0 | sort -z)
     else
       for suite_filename in "$filename"/*."${BATS_FILE_EXTENSION:-bats}"; do
-        filenames+=("$suite_filename")
+        if ! is_excluded "$suite_filename" ; then
+          filenames+=("$suite_filename")
+        fi
       done
     fi
     shopt -u nullglob
```

Comment:
It might be surprising that singly defined files are not subject to the filter. Arguably, one could expect users to not want to exclude explicitly named files. However, some users might do `bats test/*.bats` which would circumvent the `--exclude` mechanism.